### ISSUE TITLE
`object-shorthand` and `quote-props` have incompatibilities

### ISSUE THREAD

**Issue by joakimbeng** [eslint/eslint#3544]
```bash
$> eslint -v
v1.2.1
```

My `.eslintrc`:

```json
{
  "ecmaFeatures": {
    "objectLiteralShorthandProperties": true
  },
  "rules": {
    "quote-props": [2, "consistent"],
    "object-shorthand": [2, "always"]
  }
}
```

The code:

```javascript
var x = true;
({
  x,
  'needs-quote': true
});
```

Running `eslint` gives the following output:

```bash
index.js
  2:2  error  Inconsistently quoted property `needs-quote` found  quote-props
```

Changing the code to:

```javascript
var x = true;
({
  'x': x,
  'needs-quote': true
});
```

Will instead give the error:

```bash
index.js
  3:3  error  Expected property shorthand  object-shorthand
```

So there's a Catch-22 between these two rules. I could of course rename the `x` variable to something else, but I would prefer to be able to use the object property shorthand syntax even when I have `"consistent"` or `"consistent-as-needed"` for the `quote-props` rule.

What do you think? I could probably solve this myself and submit a PR but the question is what's actually the best solution for this? Or maybe you're of the opinion that this is by design and the rules should really be incompatible when using this combination of options?

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by BYK** [comment]
I think you need to drop `object-shorthand` and use `"consistent-as-needed"` option for `quote-props`.

**Comment by mysticatea** [comment]
Thank you for the report.

I think this is a bug of `quote-props`.
This rule should ignore computed/shorthand properties.

```js
/* eslint-env es6 */
/* eslint quote-props: [2, "always"] */
var x;
var obj = ({
    [x]: 0,
    x
});
```

Actual:

```
> eslint -v
v1.2.1
> eslint test.js --no-eslintrc

test.js
  5:5  error  Unquoted property `x` found  quote-props
  6:5  error  Unquoted property `x` found  quote-props

✖ 2 problems (2 errors, 0 warnings)
```

Expected:

no errors.

**Comment by BYK** [comment]
@mysticatea - he never mentioned computed properties. For short-hand, the rule does indeed check them for unnecessary quotes so I disagree with you this being a bug.

Computed shorthand part might be a bug but that needs its own ticket.

**Comment by joakimbeng** [comment]
@BYK but that's the thing, I don't want to drop the `object-shorthand` rule. I want to be able to use those strict options on both the rules at the same time without them colliding.

**Comment by BYK** [comment]
@joakimbeng - What you tell the rules to do are at odds. You are using `consistent` which says "if I ever need to quote something, everything else should be quoted too" and you also say `object-shorthand` "no matter what, make sure if I can write a property without quotes, don't quote them".

Can you tell me what you expect from the linter exactly because it seems like dropping `object-shorthand` and using `consistent-as-needed` option for `quote-props` would do precisely what you want:

- Warn about extra quotes if nothing needs them in the object definition
- Ensure everything else is quoted too if any property needs to be quoted.

**Comment by mysticatea** [comment]
@BYK Yes, @joakimbeng had not mentioned computed properties. But I check similar problems in verifying new issue. I believe it's a good practice.

In this case, computed properties is warned by this rule, it's a bug completely.

Also, I believe the shorthand properties case is a bug. That is [IdentifierReference, not PropertyName](http://www.ecma-international.org/ecma-262/6.0/#sec-object-initializer).

**Comment by mysticatea** [comment]
Ah, sorry, I will open new issue for computed properties.

**Comment by BYK** [comment]
>Also, I believe the shorthand properties case is a bug. That is IdentifierReference, not PropertyName.

Not sure what you are referring to with this :( This whole ticket is about 2 rules colliding. I am saying this is due to misconfiguration and don't understand why any of you think this is an actual bug.

**Comment by joakimbeng** [comment]
@BYK 
> you also say object-shorthand "no matter what, make sure if I can write a property without quotes, don't quote them".

That's not what that rule is for. That rule says that if you can use the object-shorthand syntax you should use it, and it says nothing about quotes.

**Comment by mysticatea** [comment]
@BYK My intent is "This rule should ignore shorthand properties."

```
/* eslint-env es6 */
/* eslint quote-props: [2, "consistent"] */
var x;
var obj1 = ({
    "a-b": 0,
    x
});
var obj2 = ({
    x,
    "a-b": 0
});
```

```
> eslint test.js --no-eslintrc

test.js
  4:13  error  Inconsistently quoted property `x` found    quote-props
  8:13  error  Inconsistently quoted property `a-b` found  quote-props

✖ 2 problems (2 errors, 0 warnings)
```

I thought this is an unexpected behavior. Because `x` is an [IdentifierReference, not PropertyName](http://www.ecma-international.org/ecma-262/6.0/#sec-object-initializer).

**Comment by joakimbeng** [comment]
I'm with @mysticatea on this, I think that `quote-props` should ignore shorthand properties.

@sindresorhus do you have any input on this? XO has this configuration for the esnext version by default...

**Comment by BYK** [comment]
Oh, boy - yeah. I was reading this all wrong for the whole time. Yeah it should ignore shorthand properties since there's no way you can use them with quotes anyways right?

**Comment by BYK** [comment]
PS: I was seeing `x: 1` for some reason and "shorthand" didn't hit me for quite a while since I'm not using ES6 regularly.

**Comment by mysticatea** [comment]
:+1: 

Oh, the fix quickly, thanks!

**Comment by BYK** [comment]
@joakimbeng - apologies again for wasting your time :(

**Comment by joakimbeng** [comment]
@BYK hehe, no worries :)
